File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.py
# 입력 시작
N, M = map(int, input().split())
r, c, direction = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]

# 상우하좌
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

# 4방향 안에 청소할게 있을 때 방향전환
need_clean = {
    0: 3,
    1: 0,
    2: 1,
    3: 2,
}
# 청소할게 없을 때 방향전환
no_clean = {
    0: 2,
    1: 3,
    2: 0,
    3: 1,
}

arr[r][c] = 2
result = 1

is_clean = False
while True:
    # 4방향 중 청소가 가능한 곳이 있었다면 바로 앞칸을 확인해본다.
    if is_clean:
        nr = r + dr[direction]
        nc = c + dc[direction]
        if 0 <= nr < N and 0 <= nc < M:
            if arr[nr][nc] == 0:
                result += 1
                arr[nr][nc] = 2
                r = nr
                c = nc

    # 청소 가능 여부 초기화
    is_clean = False
    # 4방향을 확인하며 청소할 곳이 있는지 확인
    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        if 0 <= nr < N and 0 <= nc < M:
            # 청소가 가능한지 확인
            if arr[nr][nc] == 0:
                is_clean = True
                break
    # 4방향 안에 청소할 게 있었다면
    if is_clean:
        # 반시계 90도 회전
        direction = need_clean[direction]
    # 없었다면 뒤로 한칸
    else:
        back = no_clean[direction]
        nr = r + dr[back]
        nc = c + dc[back]
        if 0 <= nr < N and 0 <= nc < M:
            if arr[nr][nc] == 1:
                break
        else:
            break
        # 위 조건문을 통과했다면 뒤로 한칸 갈 수 있는 상태
        r = nr
        c = nc
print(result)

File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    // 델타 정의
    public static int[] dr = new int[] {-1, 0, 1, 0};
    public static int[] dc = new int[] {0, 1, 0, -1};

    // 배열의 경계 확인 메서드
    public static boolean isBoundary(int nr, int nc, int N, int M) {
        return 0 <= nr && nr < N && 0 <= nc && nc < M;
    }

    // 로봇청소기 클래스 정의
    public static class Robot {
        int r;
        int c;
        int dir;
        int result = 0;

        public Robot(int r, int c, int dir) {
            this.r = r;
            this.c = c;
            this.dir = dir;
        }

        // 청소 메서드 -> 현재 칸이 청소되지 않은 경우 청소하는 메서드
        void clear(int[][] arr) {
            arr[this.r][this.c] = 2;
            this.result += 1;
        }

        // 방향전환 메서드 -> 현재 칸의 4칸 중 청소되지 않은 빈 칸이 있는 경우
        void changeDir() {
            if (this.dir == 0) {
                this.dir = 3;
            } else {
                this.dir -= 1;
            }
        }

        // 전진 메서드 -> 바라보는 방향을 기준으로 해당 칸이 청소되지 않은 빈 칸인 경우 한 칸 전진
        void front() {
            this.r = this.r + dr[this.dir];
            this.c = this.c + dc[this.dir];
        }

        // 후진 메서드 -> 청소되지 않은 빈 칸이 없는 경우 후진
        void back() {
            this.r = this.r + dr[this.dir] * (-1);
            this.c = this.c + dc[this.dir] * (-1);
        }
    }

    public static void main(String[] args) throws IOException {
//        System.setIn(new FileInputStream("input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        final int N = Integer.parseInt(st.nextToken());
        final int M = Integer.parseInt(st.nextToken());
        // 청소기 객체 생성
        StringTokenizer st2 = new StringTokenizer(br.readLine());
        Robot robot = new Robot(
                Integer.parseInt(st2.nextToken()), Integer.parseInt(st2.nextToken()),
                Integer.parseInt(st2.nextToken()));

        int[][] arr = new int[N][M];

        for (int i = 0; i < N; i++) {
            StringTokenizer st3 = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                arr[i][j] = Integer.parseInt(st3.nextToken());
            }
        }

        // 입력 후 메인 로직 구현부
        // 청소기 객체 생성
        while (true) {
            // 0. 현재 칸 파싱
            int r = robot.r;
            int c = robot.c;
            // 1. 현재 칸이 아직 청소되지 않은 경우
            if (arr[r][c] == 0) {
                robot.clear(arr);
            }
            // 2. 현재 칸의 주변 4칸 중 청소되지 않은 빈 칸이 있는지 확인
            boolean goBack = true;
            for (int i=0; i<4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                if (isBoundary(nr, nc, N ,M)) {
                    if (arr[nr][nc] == 0) {
                        goBack = false;
                        break;
                    }
                }
            }
            if (goBack) {
                robot.back();
                if (robot.r < 0 || robot.r >= N || robot.c < 0 || robot.c >= M || arr[robot.r][robot.c] == 1) {
                    break;
                }
                continue;
            }
            // 3. 현재 칸의 주변 4칸 중 청소되지 않은 빈 칸이 있는 경우
            robot.changeDir();
            int nr = robot.r + dr[robot.dir];
            int nc = robot.c + dc[robot.dir];
            if (isBoundary(nr, nc, N ,M)) {
                // 해당 칸이 청소되지 않은 칸이면 해당 칸으로 이동
                if (arr[nr][nc] == 0) {
                    robot.r = nr;
                    robot.c = nc;
                }
            }

        }
        System.out.println(robot.result);

    }
}
